src/Strategy/Erlang/Rebar3Tree.hs

```diff
@@ -0,0 +1,156 @@
+mkProjectClosure dir deps = ProjectClosureBody
+  { bodyModuleDir    = dir
+  , bodyDependencies = dependencies
+  , bodyLicenses     = []
+  }
+  where
+  dependencies = ProjectDependencies
+    { dependenciesGraph    = buildGraph deps
+    , dependenciesOptimal  = NotOptimal
+    , dependenciesComplete = NotComplete
+    }
+
+type RebarGrapher = LabeledGrapher Rebar3Dep RebarLabel
+
+data RebarLabel =
+    RebarSource Text -- location
+  deriving (Eq, Ord, Show, Generic)
+
+buildGraph :: [Rebar3Dep] -> Graphing Dependency
+buildGraph deps = run . withLabeling toDependency $ do
+  traverse_ direct deps
+  traverse_ mkEdges deps
+
+  where
+  toDependency :: Rebar3Dep -> Set RebarLabel -> Dependency
+  toDependency pkg = foldr applyLabel start
+    where 
+    applyLabel :: RebarLabel -> Dependency -> Dependency
+    applyLabel (RebarSource src) dep = dep { dependencyLocations = src : dependencyLocations dep }
+
+    start =
+      Dependency { dependencyType =
+                      case T.isInfixOf (T.pack "github.com") (depLocation pkg) of
+                      True -> GitType
+                      False -> HexType
+                 , dependencyName =
+                      case T.isInfixOf (T.pack "github.com") (depLocation pkg) of
+                      True -> depLocation pkg
+                      False -> depName pkg
+                 , dependencyVersion = Just (CEq (depVersion pkg))
+                 , dependencyLocations = []
+                 , dependencyEnvironments = []
+                 , dependencyTags = M.empty
+                 }
+
+  mkEdges :: Has RebarGrapher sig m => Rebar3Dep -> m ()
+  mkEdges parentDep =
+    forM_ (subDeps parentDep) $ \childDep -> do
+      edge (parentDep) (childDep)
+      mkEdges childDep
+
+data Rebar3Dep = Rebar3Dep
+  { depName     :: Text
+  , depVersion  :: Text
+  , depLocation :: Text
+  , subDeps     :: [Rebar3Dep]
+  } deriving (Eq, Ord, Show, Generic)
+
+type Parser = Parsec Void Text
+
+rebar3TreeParser :: Parser [Rebar3Dep]
+rebar3TreeParser = concat <$> ((try (rebarDep 0) <|> ignoredLine) `sepBy` eol) <* eof
+  where
+  isEndLine :: Char -> Bool
+  isEndLine '\n' = True
+  isEndLine '\r' = True
+  isEndLine _    = False
+
+  -- ignore content until the end of the line
+  ignored :: Parser ()
+  ignored = do
+        i <- takeWhileP (Just "ignored") (not . isEndLine)
+        _ <- traceM $ show i
+        pure ()
+
+  ignoredLine :: Parser [Rebar3Dep]
+  ignoredLine = do
+    ignored
+    pure []
+
+  findName :: Parser Text
+  findName = takeWhileP (Just "dep") (/= '─')
+
+  findVersion :: Parser Text
+  findVersion = takeWhileP (Just "version") (/= ' ')
+
+  findLocation :: Parser Text
+  findLocation = takeWhileP (Just "location") (/= ')')
+
+  rebarDep :: Integer -> Parser [Rebar3Dep]
+  rebarDep depth = do
+    _ <- chunk " "
+    countSlash <- many "  │"
+    _ <- satisfy (\_ -> (fromIntegral (length countSlash)) == depth)
```

Comment:
this is clever! this also has another name from `Control.Monad`: `guard :: Alternative f => Bool -> f ()`:
```hs
guard (length countSlash == depth)
```

you've used `Alternative` elsewhere (it defines `<|>` and `empty`), and `guard` is defined as:
```hs
guard :: Alternative f => Bool -> f ()
guard b = if b then pure () else empty
```

Reply:
So I can't seem to get this to work, and I believe the reason has to do with how `satisfy` fails the parser if the condition is not met. `guard` seems to be causing the parser to ignore the initial `concat` in `rebar3TreeParser = concat <$> ((try (rebarDep 0) <|> ignoredLine) `sepBy` eol) <* eof` when I use it like this `_ <- guard (length slashCount == depth)`. Do you have any idea why this would be happening?

Reply:
I've also tried `guard $ not ....` but that doesn't seem to work. `satisfy` will backtrack the parser which I would expect `guard` to do as well but for some reason it is not.